### DISCUSSION TITLE
multiblock performance

### DISCUSSION THREAD

**Discussion by Ethanedc** [MartinTheDragon/Nuclear-Tech-Mod-Remake#59] [General]
The forge API can let you render blocks or something yourself?
like generate the buffers, store the data, choose the shaders, yourself, without the rendering system of minecraft

**Comment by MartinTheDragon** [comment]
Hello!
I can't quite understand what you're trying to say.
Is this an issue? Is this a question? Is this supposed to give me information?
The easiest to use thing Forge provides is `SimpleRenderable` as a sort of wrapper to vanilla model rendering. You can get an instance via `OBJModel#bakeRenderable`.
In the end it still pipes into Mojang's Blaze3D, which relies on OpenGL and GLFW, uses two frame buffers, several different vertex buffers and uploads and renders vertex data multiple times in one frame via `glBufferData` and `glDrawElements` respectively, which is pretty expensive.
I'll probably attempt GPU instancing at some point by keeping track of render positions and transformations myself. Though it will require even more custom shaders and probably break a lot of shader packs.

**Comment by Ethanedc** [comment]
In fact that is my question, if the OpenGL and GLFW API could be used internally
because i don't know how works the forge API and i will help with the multiblock performance if i can, because I have already worked with the API before(a little and in c++)

**Comment by coaber** [comment]
The GPU instancing could perhaps be a better option and it actually outweighs the high chance of breaking shaders, performance plays a very big part in even the smallest modpacks, and building large multiblocks like the RBMK can cause the performance to take a significant hit when used in a large modpack.
